### PR TITLE
Skip empty hooks

### PR DESCRIPTION
**Pull request by damianszczepanik**
Issue #841

### CODE REVIEW

File: src/main/java/net/masterthought/cucumber/json/Hook.java [hunk 1]
```diff
@@ -38,4 +39,21 @@ public Match getMatch() {
...
+     */
+    public boolean hasContent() {
+        if (embeddings.length > 0) {
+            // assuming that if the embedding exists then it is not empty
+            return true;
+        }
+        if (StringUtils.isNotBlank(result.getErrorMessage())) {
+            return true;
+        }
+        // TODO: hook with 'output' should be treated as empty or not?
```
Comment (KLBonn) [inline src/main/java/net/masterthought/cucumber/json/Hook.java]: IMO this implementation is OK. I have not seen important output from hooks (aside from embeddings / exception error messages) yet.
If anyone should miss further output from hooks, she can simply disable the hide_empty_hooks feature.
So as far as I am concerned, feel free to remove TODO comment.

File: src/main/java/net/masterthought/cucumber/util/Util.java [hunk 2]
```diff
@@ -76,4 +81,16 @@ public static String toValidFileName(String fileName) {
...
         return Long.toString((long) fileName.hashCode() + Integer.MAX_VALUE);
     }
+
+    /**
+     * Helper method that removes empty hooks from passed array and packs it into new collection.
+     *
+     * @param hooks hooks to be reduced
+     * @return no empty hooks
+     */
+    public static List<Hook> eliminateEmptyHooks(Hook[] hooks) {
```
Comment (KLBonn) [inline src/main/java/net/masterthought/cucumber/util/Util.java]: I am not familiar with the template engine. Could the different types cause problems? You get an array, and return a java.util.List object.
In hooks.vm below you _update_ $hooks with the method result, so you change its type. Do the macros support this?

File: src/main/resources/templates/macros/json/hooks.vm [hunk 3]
```diff
@@ -1,5 +1,9 @@
 #macro(includeHooks $keyword, $hooks, $status, $level)
 
+#if($hide_empty_hooks)
+  #set($hooks = $util.eliminateEmptyHooks($hooks))
+#end
+
 #if(!$hooks.isEmpty())
```
Comment (KLBonn) [inline src/main/resources/templates/macros/json/hooks.vm]: See type-change comment above at Util.java: 
If NOT $hide_empty_hooks, then $hooks still is of type array, isn't it? And array does not have method .isEmpty(), I think?
Or is there some kind of runtime conversion?
Not sure if this is a problem.

File: src/main/java/net/masterthought/cucumber/util/Util.java (hunk 2 above)
Comment (damianszczepanik) [reply src/main/java/net/masterthought/cucumber/util/Util.java]: it converts smoothly between array and collections

File: src/main/resources/templates/macros/json/hooks.vm (hunk 3 above)
Comment (damianszczepanik) [reply src/main/resources/templates/macros/json/hooks.vm]: no flag -> stays unchanged as it is right now

isEmpty is the method of velocity template, not pure java in this case

**Comment by damianszczepanik** [comment]
I've tested the change against JSON you provided and hooks were not there at all (except a couple that have some value AFAIR)

**Comment by KLBonn** [comment]
OK, then it looks simply good to me :+1:

**Pull request merged**